Improve stopping

Martchus opened the pull request:
* See particular commit messages
* See https://progress.opensuse.org/issues/67342 for the full story
    * TL;DR: We want to allow ffmpeg to finalize its video.
        * For that we need to prevent spamming SIGTERM signals on it without introducing long sleep times.
        * We also need to be able to stop the full process group and wait for the termination of the full process group to know when ffmpeg has finalized the video.

okurz commented on lib/Mojo/IOLoop/ReadWriteProcess.pm:
```diff
@@ -32,7 +33,7 @@ use Exporter 'import';
 
 has [
-  qw(kill_sleeptime sleeptime_during_kill),
+  qw(kill_sleeptime sleeptime_during_kill patience_during_kill),
```
If you want to introduce another parameter please include in documentation. "patience" and "sleeptime" are IMHO ambiguous in understanding. How about "interval_time" or "total_sleeptime_during_kill"?

Martchus replied:
I suppose `total_sleeptime_during_kill` makes sense.

okurz commented on lib/Mojo/IOLoop/ReadWriteProcess.pm:
```diff
@@ -46,6 +47,7 @@ has [
 ];
 
 has max_kill_attempts => 5;
+has kill_whole_group  => 0;
```
that deserves some documentation as well, e.g. the text from commit message

foursixnine replied:
There's a section in the POD where it should be added

Martchus replied:
I have already pushed a change to extend the POD. Did I add it to the wrong place? By the way, I've tried to be consistent with the existing "style".

foursixnine replied:
It was in the right place :)

Martchus commented:
I would also like to note that I haven't had that much time yesterday to test it. So I'll better add a WIP prefix.

Martchus commented:
I was on vacation and now the CI tests on openQA are broken so I haven't had the time to look into this PR further (but I have not abandoned it). The suggestions so far make sense and I'll of course look into the failing tests.

foursixnine commented:
@Martchus I have restarted the failing jobs, once they pass, I think we can move forward

Martchus commented:
I've now updated the code and added/adapted unit tests. It seems to work now. I guess the failing tests for Perl < 5.16 can be ignored?

By the way, I've also tested this by running and stopping an openQA job using ffmpeg as videoencoder. Everything seems to work (no delays, no truncated video including the uploaded video).

foursixnine commented:
@Martchus Can you rebase?

Martchus commented:
Rebased. By the way, it would be nice to make a new release once this change has been merged so we can make use of it in openQA without making a Git package.

Martchus commented:
I've also realized that with the latest version of tidy there's a huge diff. So maybe before the release I can create yet another PR to simply apply the latest version of tidy.

(The changes here are already tidied with the latest version of tidy but I didn't include unrelated tidying.)

foursixnine commented:
> I've also realized that with the latest version of tidy there's a huge diff. So maybe before the release I can create yet another PR to simply apply the latest version of tidy.
> 
> (The changes here are already tidied with the latest version of tidy but I didn't include unrelated tidying.)

@mudler usually does the releases, so pre or post tidy would be more up to him, but sending the PR might make sense
